### PR TITLE
Add recent fulfilers option group through integration with Customizr

### PR DESCRIPTION
**Pull request by Igrom**
Secondary points:
- the code uses async/await. In order to use it, I needed a polyfill for generator functions: https://github.com/facebook/regenerator. See how it's included as an entry point in the webpack config.
- set up Storybook for more convenient development
- created a simple Customizr client
- changed componentWillReceiveProps to componentDidUpdate to appease a React method deprecation
- added new strings to Translations (note: I think they have to be added to Bulgarian first to show in other langauges).

Ugly things:
- added aws-sdk as a dev dependency. There's an issue for this: https://github.com/Cimpress/cimpress-translations-webpack-plugin/issues/4
- there's probably a better way of swapping connection strings than checking global scope (FulfillerSelect.jsx)
- cimpress-fulfiller-identity enforces HTTP. Which means no http://localhost. A Storybook story (loading fulfillers from the mock) fails because of it.

---

Main point: reengineered the option rendering code and split the drop-down into two option groups.

screenshot from 2018-06-25 07-11-08

Initial idea: use react-select v2 beta, which includes option group support, like so:
```
options: [
    {
        label: 'abc',
        options: [1,2,3]
    },
    {
        label: 'def',
        options: [4,5,6]
    }
]
```
Obstacle: our list is big --- there's several hundred fulfillers. `react-select` renders options simplistically, most likely by adding all options to the DOM at once, which is inefficient at higher option counts. It is better to use virtual scrolling, which AFAIK dynamically adds and removes nodes as you scroll. For this reason we have been using `react-virtualized-select`, which is a `react-select` wrapper that replaces the default `menuRenderer` method with its own implementation. Because of this, I could not use `react-select` v2.

Solution: if I can't use option groups, I can just render option group labels as options that are disabled, unfocusable, unsearchable in the free-form text query and differently styled. There's not much they share with regular options afterwards.

There's some polymorphism going on with option rendering. `getOptions()` returns objects which have an $.option.optionRenderer method property, which knows how to render that particular component (spinner "Loading...", warning/error message, normal option, group label).

Whenever you select an option:
- a fresh list of recent fulfillers (IDs) is retrieved from Customizr
- your pick is added to the list. If it was already there, it bubbles up to the head of the list.
- the list is saved.

While the component limits how many fulfillers are shown, there's no limit to what's saved to Customizr (a UI component shouldn't make such decisions). It also does not show archived fulfillers in the "recent" group. What do you think about that?

Cues for writing an option renderer were taken from https://github.com/bvaughn/react-virtualized-select/blob/master/source/VirtualizedSelect/VirtualizedSelect.js.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -20,15 +21,20 @@
...
-    "@cimpress/react-components": "^3.3.4",
+    "@cimpress/react-components": "^4.0.0",
+    "axios": "^0.18.0",
     "cimpress-fulfiller-identity": "^0.1.4",
     "cimpress-fulfillment-location": "^0.1.10",
     "i18next": "^11.3.2",
     "react-highlight-words": "^0.10.0",
     "react-i18next": "^7.7.0",
-    "react-virtualized-select": "^3.1.3"
+    "react-select": "^2.0.0-beta.6",
```
[the remark this answers is not in the document]
Comment (Igrom) [reply package.json]: I need to downgrade it. Ultimately we're not using it.

File: package.json [hunk 2]
```diff
@@ -43,10 +49,13 @@
...
     "load-grunt-tasks": "^3.5.2",
-    "react": "^15.0.0 || ^16.0.0",
-    "react-dom": "^15.0.0 || ^16.0.0",
+    "mockserver": "^1.11.0",
+    "nock": "^9.3.3",
+    "react": "^16.0.0",
+    "react-dom": "^16.0.0",
     "style-loader": "^0.21.0",
-    "webpack": "^4.11.0"
+    "webpack": "^3.0.0"
```
[the remark this answers is not in the document]
Comment (Igrom) [reply package.json]: I could not get Storybook to work under v4. If anyone knows how to do it, they're welcome to try.

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (Igrom) [reply package.json]: Downgraded.

File: src/FulfillerSelect.jsx [hunk 3]
```diff
@@ -1,9 +1,12 @@
 import React from 'react';
 import PropTypes from 'prop-types';
-import Select from 'react-virtualized-select';
+import VirtualizedSelect from 'react-virtualized-select';
 import SelectWrapper from '@cimpress/react-components/lib/SelectWrapper';
 import {shapes} from '@cimpress/react-components';
 import {getFulfillers} from './apis/fi.api';
+import CustomizrClient from './apis/customizr.api';
+
+import '../styles/index.css'
```
[the remark this answers is not in the document]
Comment (Igrom) [reply src/FulfillerSelect.jsx]: Changed.

File: src/apis/customizr.api.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
+const { pope } = require('pope');
+const axios = require('axios');
+
+const DEFAULT_BASE_URL = 'https://customizr.at.cipress.io';
+const API = {
+  v1Settings: '/v1/resources/{{resource}}/settings'
+};
+
+class CustomizrClient {
```
[the remark this answers is not in the document]
Comment (Igrom) [reply src/apis/customizr.api.js]: Long overdue. As mentioned, AT Squad has repeatedly stated that they won't write clients to some of their services. Let's do it ourselves.

File: src/FulfillerSelect.jsx [hunk 5]
```diff
@@ -18,8 +21,19 @@ class FulfillerSelect extends React.Component {
...
+            selectedFulfillerId: undefined,
+            recentFulfillerIds: [],
+            fetchingFulfillers: true
         };
+
+        this.customizrClient = new CustomizrClient(global.CUSTOMIZR_URL || null, 'https://trdlnk.cimpress.io');
+
+        this.onChange = this.onChange.bind(this);
+        this.fulfillerOptionGroupLabelOptionRenderer = this.fulfillerOptionGroupLabelOptionRenderer.bind(this);
+        this.fulfillerSingleOptionRenderer = this.fulfillerSingleOptionRenderer.bind(this);
```
[the remark this answers is not in the document]
Comment (Igrom) [reply src/FulfillerSelect.jsx]: Done.

[the remark this answers is not in the document]
Comment (Igrom) [reply src/FulfillerSelect.jsx]: Actually, I take it back. If you'd like to sell the code, then options => this.fulfiller<renderer name>Renderer(options) arrow functions sprinkled throughout the code are unappealing.

I might change it if I address the other comment (not using custom renderers)

[the remark this answers is not in the document]
Comment (istanishev) [reply src/FulfillerSelect.jsx]: It's always funny how subjective are these things, I've seen exactly the opposite opinions before  :P

File: src/apis/customizr.api.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (istanishev) [reply src/apis/customizr.api.js]: 👍

File: src/FulfillerSelect.jsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (Igrom) [reply src/FulfillerSelect.jsx]: Is it really preferable (according to general opinion) to have arrow functions? If so, I can change it.

File: src/FulfillerSelect.jsx [hunk 6]
```diff
@@ -43,73 +57,203 @@ class FulfillerSelect extends React.Component {
...
-                label: this.getTitle(f)
+        let fulfillerOptions = fulfillers.map(f => Object.assign({}, f, {
+            optionRenderer: this.fulfillerSingleOptionRenderer,
+            value: `${f.fulfillerId} ${f.name}` // 'value' required for search and focus style change functionality
+        }));
+
+        let recentFulfillerOptions = this.state.recentFulfillerIds
+            .slice(0, 5)
+            .map(id => fulfillerOptions.find(f => f.fulfillerId === id))
+            .filter(f => f && !f.archived)
```
[the remark this answers is not in the document]
Comment (Igrom) [reply src/FulfillerSelect.jsx]: Done!

File: package.json [hunk 7]
```diff
@@ -20,15 +21,20 @@
     "react": "^15.0.0 || ^16.0.0"
   },
   "dependencies": {
-    "@cimpress/react-components": "^3.3.4",
+    "@cimpress/react-components": "^4.0.0",
```
[the remark this answers is not in the document]
Comment (Igrom) [reply package.json]: Downgraded.

File: src/FulfillerSelect.jsx [hunk 8]
```diff
@@ -43,73 +57,203 @@ class FulfillerSelect extends React.Component {
...
+            .map((f, index) => Object.assign({}, f, { value: `recent${index}` })); // do not highlight or show in search results
+
+        let recentFulfillersOptionGroupLabelOption = {
+            text: this.tt('recent-fulfillers'),
+            optionRenderer: this.fulfillerOptionGroupLabelOptionRenderer
```
[the remark this answers is not in the document]
Comment (Igrom) [reply src/FulfillerSelect.jsx]: Closing today's standup discussion, let's revisit this if a need for it resurfaces. It is likely that with the impending release of react-select v2, other changes will trump these.

**Pull request merged**